kernel/aster-nix/src/time/wait.rs

```diff
@@ -1,27 +1,96 @@
 // SPDX-License-Identifier: MPL-2.0
 
-use core::time::Duration;
+use core::{
+    sync::atomic::{AtomicBool, Ordering},
+    time::Duration,
+};
 
-use ostd::sync::{WaitQueue, Waiter};
+use ostd::sync::{WaitQueue, Waiter, Waker};
 
-use super::{clocks::JIFFIES_TIMER_MANAGER, timer::Timeout};
+use super::{clocks::JIFFIES_TIMER_MANAGER, timer::Timeout, Timer, TimerManager};
+use crate::prelude::*;
 
 /// A trait that provide the timeout related function for [`WaitQueue`]`.
 pub trait WaitTimeout {
     /// Wait until some condition returns `Some(_)`, or a given timeout is reached. If
     /// the condition does not becomes `Some(_)` before the timeout is reached, the
     /// function will return `None`.
+    ///
+    /// The timeout is against the default JIFFIES clock.
     fn wait_until_or_timeout<F, R>(&self, cond: F, timeout: &Duration) -> Option<R>
+    where
+        F: FnMut() -> Option<R>,
+    {
+        let timeout_against_clock = WakerTimerCreater::new(timeout);
+        self.wait_until_or_timeout_against_clock(cond, &timeout_against_clock)
+    }
+
+    /// Similar to [`WaitTimeout::wait_until_or_timeout`].
+    ///
+    /// The difference is that the timeout of this method is against the specified clock,
+    /// which is defined in `timer_creater`.
+    fn wait_until_or_timeout_against_clock<F, R>(
+        &self,
+        cond: F,
+        timer_creater: &WakerTimerCreater,
+    ) -> Option<R>
     where
         F: FnMut() -> Option<R>;
 }
 
+/// A struct for creating [`Timer`] against a specific timer manager for [`Waker`].
+pub struct WakerTimerCreater<'a> {
```

Comment:
I agree with the idea that something like `WakerTimerCreater` should be added. But its name and interface is a bit of confusing. This `WakerTimerCreater` name reads like three objects (similar to `RockPaperCissors`), not one. And from its public interface, the user cannot see why it has anything to do with waker or timer or creator: its constructor method does not take a waker or timer argument; its getter or setter method does not take a waker or timer argument; it has no method that can _create_ a timer. Overall, I think it is not a good abstraction.

Consider the following revised form.

```rust
/// A helper structure to build timers from a specified timeout and timer manager.
pub struct TimerBuilder<'a> {
    timeout: Timeout,
    timer_manager: &'a TimerManager,
}

impl<'a> TimerBuilder<'a> {
    pub fn new(timeout: &Duration) -> Self {
        todo!()
    }

    pub fn new_with_timer(timeout: Timeout, timer_manager: &'a TimerManager) -> Self {
        todo!()
    }

    pub fn timeout(&self) -> &Timeout {
        &self.timeout
    }

    pub fn timer_manager(&self) -> &TimerManager {
        &self.timer_manager
    }

    /// Builds and sets a timer,
    /// which will trigger `callback` when `self.timeout()` is reached.
    pub fn fire(&self, callback: F) -> Arc<Timer>
    where
        F: Fn() + Send + Sync + 'static,
    {
        todo!()
    }
}

pub trait WaitTimeoutUnconditionl {
    fn wait_or_timeout(&self, timeout: &Duration) -> Result<()> {
        self.wait_or_timeout_against_clock(&TimerBuilder::new(timeout))
    }

    fn wait_or_timer_timeout(&self, timer_builder: &TimerBuilder) -> Result<()>;
}


```